### PR TITLE
[AMDGPU][SIInsertWaitcnts] Do not add s_waitcnt when the counters are known to be 0 already

### PR DESCRIPTION
**Pull request by jmmartinez**
Currently a WIP, I've only updated a single test to depict the change (there are ~50 tests that need update).

Differential Revision: https://reviews.llvm.org/D156679

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 1]
```diff
@@ -869,6 +876,26 @@ static bool updateOperandIfDifferent(MachineInstr &MI, uint16_t OpName,
...
 
+bool SIInsertWaitcnts::updateWaitcntIfSoft(MachineInstr *&Waitcnt) const {
+  unsigned Opcode = Waitcnt->getOpcode();
+  if (!SIInstrInfo::isSoftWaitcnt(Opcode))
+    return false;
+
+  MachineBasicBlock &Block = *Waitcnt->getParent();
+  auto InsertBefore = Waitcnt->getIterator();
+  auto DL = Waitcnt->getDebugLoc();
+  auto &MCInfo = TII->get(SIInstrInfo::getNonSoftWaitcntOpcode(Opcode));
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:887]: You can just setDesc instead of creating a whole new instruction

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 2]
```diff
@@ -477,7 +483,7 @@ class SIInsertWaitcnts : public MachineFunctionPass {
   bool generateWaitcnt(AMDGPU::Waitcnt Wait,
                        MachineBasicBlock::instr_iterator It,
                        MachineBasicBlock &Block, WaitcntBrackets &ScoreBrackets,
-                       MachineInstr *OldWaitcntInstr);
+                       MachineInstr *OldWaitcntInstr) const;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:486]: Looks like an unrelated clean up? Please commit it separately.

File: llvm/test/CodeGen/AMDGPU/back-off-barrier-subtarget-feature.ll [hunk 3]
```diff
@@ -45,6 +45,7 @@ define void @back_off_barrier_no_fence(ptr %in, ptr %out) #0 {
 ; GFX11-BACKOFF-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
 ; GFX11-BACKOFF-NEXT:    flat_load_b32 v0, v[0:1]
 ; GFX11-BACKOFF-NEXT:    s_waitcnt vmcnt(0) lgkmcnt(0)
+; GFX11-BACKOFF-NEXT:    s_waitcnt_vscnt null, 0x0
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/back-off-barrier-subtarget-feature.ll:48]: I don't understand why your patch would ever cause an extra waitcnt like this.

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 4]
```diff
@@ -364,7 +371,6 @@ class SIInsertWaitcnts : public MachineFunctionPass {
   const MachineRegisterInfo *MRI = nullptr;
   AMDGPU::IsaVersion IV;
 
-  DenseSet<MachineInstr *> TrackedWaitcntSet;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:367]: Is my understanding correct? Previously we used TrackedWaitcntSet to distinguish pre-existing waitcnts from waitcnts inserted by this pass. Now we don't need that, all we need it to know the difference between hard and soft waitcnts?

File: llvm/lib/Target/AMDGPU/SOPInstructions.td [hunk 5]
```diff
@@ -1277,9 +1277,18 @@ def S_WAKEUP : SOPP_Pseudo <"s_wakeup", (ins) > {
   let mayStore = 1;
 }
 
-let hasSideEffects = 1 in
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SOPInstructions.td:1280]: Why is this removed?

File: llvm/lib/Target/AMDGPU/SOPInstructions.td [hunk 6]
```diff
@@ -1277,9 +1277,18 @@ def S_WAKEUP : SOPP_Pseudo <"s_wakeup", (ins) > {
...
     [(int_amdgcn_s_waitcnt timm:$simm16)]>;
+
+// "_soft" waitcnts are waitcnts whose wait can be relaxed or completely removed.
+// These are inserted by to resolve memory dependencies by the memory legalizer and later optimized by SIInsertWaitcnts
+// For example, a S_WAITCNT_soft 0 can be completely removed on a function that doesn't access memory.
+def S_WAITCNT_soft : SOPP_Pseudo <"s_soft_waitcnt" , (ins SWaitCnt:$simm16), "$simm16">;
+def S_WAITCNT_VSCNT_soft : SOPP_Pseudo<"s_soft_waitcnt_vscnt", (ins SReg_32:$sdst, s16imm:$simm16), "$sdst, $simm16"> {
+  let mayLoad = 1;
+  let mayStore = 1;
+  let has_sdst = 1;
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SOPInstructions.td:1290]: Why are these needed? The definition of S_WAITCNT above does not have them.

**Comment by jayfoad** [comment]
1st, 2nd and 4th commits seem like obvious cleanups. LGTM, please commit them separately.

**Comment by kzhuravl** [comment]
Hi Juan, the CQE cycle came back green, so this PR can be merged. Thanks!

**Comment by jmmartinez** [comment]
@kzhuravl that's awesome! Thanks for the update.

However, I am not able to rebase and address Jay's review since I don't have a PC available (for the next 6 months).

If somebody else could move the PR forward it would be great. Otherwise I'll finish it when I'm back.

**Comment by kzhuravl** [comment]
Hi @jwanggit86, can you please take over this PR? This needs to be rebased, and review comments addressed, thanks!

**Comment by jwanggit86** [comment]
@kzhuravl Sure. I'll take over.

**Comment by kzhuravl** [comment]
@Pierre-vh got some cycles to look into this now, so re-assigning to him.

**Comment by Pierre-vh** [comment]
#72830
